File: backend/app/models/agent.py
from pydantic import BaseModel, HttpUrl, Field
from typing import Optional
from enum import Enum
from datetime import datetime
import uuid


class PricingModel(str, Enum):
    subscription = "subscription"
    usage_based = "usage_based"


class TrackRecord(BaseModel):
    total_processed: int = Field(ge=0)
    uptime_percentage: float = Field(ge=0, le=100)
    avg_response_ms: float = Field(ge=0)
    error_rate: float = Field(ge=0, le=100)
    last_active_at: Optional[datetime] = None


class AgentBase(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    avatar_url: Optional[HttpUrl] = None
    personality: Optional[str] = Field(None, max_length=500)
    skills: list[str] = []
    track_record: Optional[TrackRecord] = None
    pricing_model: PricingModel
    api_endpoint: Optional[HttpUrl] = None


class AgentCreate(AgentBase):
    pass


class AgentUpdate(AgentBase):
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    pricing_model: Optional[PricingModel] = None


class Agent(AgentBase):
    id: uuid.UUID
    developer_id: uuid.UUID
    is_verified: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ResumeGenerateRequest(BaseModel):
    github_url: Optional[HttpUrl] = None
    api_doc_url: Optional[HttpUrl] = None

    def model_post_init(self, __context):
        if not self.github_url and not self.api_doc_url:
            raise ValueError("github_url か api_doc_url のいずれかが必要です")

File: src/schemas/agent.ts
import { z } from "zod";

export const PricingModelSchema = z.enum(["subscription", "usage_based"]);
export type PricingModel = z.infer<typeof PricingModelSchema>;

export const TrackRecordSchema = z.object({
  total_processed: z.number().int().nonnegative(),
  uptime_percentage: z.number().min(0).max(100),
  avg_response_ms: z.number().nonnegative(),
  error_rate: z.number().min(0).max(100),
  last_active_at: z.string().datetime().nullable(),
});
export type TrackRecord = z.infer<typeof TrackRecordSchema>;

export const AgentSchema = z.object({
  id: z.string().uuid(),
  developer_id: z.string().uuid(),
  name: z.string().min(1).max(100),
  avatar_url: z.string().url().nullable(),
  personality: z.string().max(500).nullable(),
  skills: z.array(z.string()),
  track_record: TrackRecordSchema.nullable(),
  pricing_model: PricingModelSchema,
  api_endpoint: z.string().url().nullable(),
  is_verified: z.boolean(),
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
});
export type Agent = z.infer<typeof AgentSchema>;

export const CreateAgentSchema = AgentSchema.omit({
  id: true,
  developer_id: true,
  is_verified: true,
  created_at: true,
  updated_at: true,
});
export type CreateAgentInput = z.infer<typeof CreateAgentSchema>;

export const UpdateAgentSchema = CreateAgentSchema.partial();
export type UpdateAgentInput = z.infer<typeof UpdateAgentSchema>;

export const ResumeGenerateSchema = z.object({
  github_url: z.string().url().optional(),
  api_doc_url: z.string().url().optional(),
}).refine(
  (data) => data.github_url || data.api_doc_url,
  { message: "github_url か api_doc_url のいずれかが必要です" }
);
export type ResumeGenerateInput = z.infer<typeof ResumeGenerateSchema>;
